javav2/example_code/s3/src/main/java/com/example/s3/express/S3DirectoriesActions.java

```diff
@@ -0,0 +1,551 @@
+                }
+                throw new CompletionException("Failed to get the object", exception);
+            })
+            .thenApply(response -> {
+                logger.info("Successfully obtained bytes from an S3 object");
+                return response;
+            });
+    }
+    /**
+     * Asynchronously copies an object from one S3 bucket to another.
+     *
+     * @param s3Client           the S3 async client to use for the copy operation
+     * @param sourceBucket       the name of the source bucket
+     * @param sourceKey          the key of the object to be copied in the source bucket
+     * @param destinationBucket  the name of the destination bucket
+     * @param destinationKey     the key of the copied object in the destination bucket
+     * @return a {@link CompletableFuture} that completes when the copy operation is finished
+     */
+    public CompletableFuture<Void> copyObjectAsync(S3AsyncClient s3Client, String sourceBucket, String sourceKey, String destinationBucket, String destinationKey) {
+        CopyObjectRequest copyRequest = CopyObjectRequest.builder()
+            .sourceBucket(sourceBucket)
+            .sourceKey(sourceKey)
+            .destinationBucket(destinationBucket)
+            .destinationKey(destinationKey)
+            .build();
+
+        return s3Client.copyObject(copyRequest)
+            .thenRun(() -> logger.info("Copied object '" + sourceKey + "' from bucket '" + sourceBucket + "' to bucket '" + destinationBucket + "'"))
+            .whenComplete((ignored, exception) -> {
+                if (exception != null) {
+                    Throwable cause = exception.getCause();
+                    if (cause instanceof S3Exception) {
+                        throw new CompletionException("Couldn't copy object '" + sourceKey + "' from bucket '" + sourceBucket + "' to bucket '" + destinationBucket + "'. Reason: " + ((S3Exception) cause).awsErrorDetails().errorMessage(), cause);
+                    }
+                    throw new CompletionException("Failed to copy object", exception);
+                }
+            });
+    }
+
+    /**
+     * Creates an asynchronous session for the specified S3 bucket.
+     *
+     * @param s3Client the S3 asynchronous client to use for creating the session
+     * @param bucketName the name of the S3 bucket for which to create the session
+     * @return a {@link CompletableFuture} that completes when the session is created, or throws a {@link CompletionException} if an error occurs
+     */
+    public CompletableFuture<Void> createSessionAsync(S3AsyncClient s3Client, String bucketName) {
+        CreateSessionRequest request = CreateSessionRequest.builder()
+            .bucket(bucketName)
+            .build();
+
+        return s3Client.createSession(request)
+            .thenRun(() -> logger.info("Created session for bucket: " + bucketName))
```

Comment:
Why not do the logging in the `whenComplete`?

Reply:
Fixed